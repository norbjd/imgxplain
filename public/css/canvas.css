#canvases {
  position: relative;
}

#canvases canvas {
  position: absolute;
  top: 0;
  border: solid black;
}

#draw {
  z-index: 1;
}

#preview {
  z-index: 2;
}
