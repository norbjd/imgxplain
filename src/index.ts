import { App } from "./app";

window.addEventListener("DOMContentLoaded", () => {
  if (!window.AudioContext) {
    const errorMessage = "Web Audio not supported, change your browser";
    alert(errorMessage);
    throw new Error(errorMessage);
  }

  new App().run();
});
