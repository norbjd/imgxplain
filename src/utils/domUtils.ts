class DOMUtils {
  static getDocumentWidth(): number {
    return Math.max(
      document.body.scrollWidth,
      document.documentElement.scrollWidth,
      document.body.offsetWidth,
      document.documentElement.offsetWidth,
      document.documentElement.clientWidth
    );
  }

  static getDocumentHeight(): number {
    return Math.max(
      document.body.scrollHeight,
      document.documentElement.scrollHeight,
      document.body.offsetHeight,
      document.documentElement.offsetHeight,
      document.documentElement.clientHeight
    );
  }

  static getTranscriptDiv(): HTMLDivElement {
    return document.querySelector("#transcript") as HTMLDivElement;
  }

  static getMainCanvas(): HTMLCanvasElement {
    return document.querySelector("#draw") as HTMLCanvasElement;
  }

  static getPreviewCanvas(): HTMLCanvasElement {
    return document.querySelector("#preview") as HTMLCanvasElement;
  }

  static getMainDiv(): HTMLDivElement {
    return document.querySelector("#main") as HTMLDivElement;
  }

  static getWaveformDiv(): HTMLDivElement {
    return document.querySelector("#waveform");
  }

  static getEditorDiv(): HTMLDivElement {
    return document.querySelector("#editor");
  }

  static getEditForm(): HTMLFormElement {
    return document.forms.namedItem("edit");
  }

  static getLoadForm(): HTMLFormElement {
    return document.forms.namedItem("load");
  }

  static getAvailableMaxWidthForMainDiv(): number {
    return (
      (this.getDocumentWidth() -
        this.getMainDivMarginLeft() -
        this.getEditorDivWidthInPx()) *
      0.95
    );
  }

  static getAvailableMaxHeightForMainDiv(): number {
    return this.getDocumentHeight() - 60;
  }

  static getEditorDivWidthInPx(): number {
    const editorDivWidth = this.getEditorDiv()
      .style.width.replace("%", "")
      .replace("px", "");
    if (editorDivWidth == "0" || editorDivWidth == "") {
      return 0;
    }
    return (this.getDocumentWidth() * parseInt(editorDivWidth)) / 100 + 20;
  }

  static getMainDivMarginLeft(): number {
    return Number(this.getMainDiv().style.marginLeft.replace("px", ""));
  }

  static setHTMLInputElementValue(
    elementName: string,
    value: string | number
  ): void {
    (document.getElementById(elementName) as HTMLInputElement).value =
      "" + value;
  }

  static getPlayPauseButton(): HTMLButtonElement {
    return document.querySelector("[data-action='play']") as HTMLButtonElement;
  }

  static getTranscriptButton(): HTMLInputElement {
    return document.querySelector("#transcript_button");
  }

  static getCCButton(): HTMLInputElement {
    return document.querySelector("#cc");
  }

  static getWaveformZoomSlider(): HTMLInputElement {
    return document.querySelector("[data-action='zoom']");
  }

  static getExportActionsToJsonButton(): HTMLButtonElement {
    return document.querySelector(
      "[data-action='export']"
    ) as HTMLButtonElement;
  }

  static getExportToVideoButton(): HTMLButtonElement {
    return document.querySelector(
      "[data-action='export-video']"
    ) as HTMLButtonElement;
  }

  static getToggleEditorButton(): HTMLButtonElement {
    return document.getElementById("toggle_editor_button") as HTMLButtonElement;
  }

  static openEditor(): void {
    this.getToggleEditorButton().textContent = "ðŸ–‰ Close Editor";
    DOMUtils.getEditorDiv().style.width = "40%";
    DOMUtils.getEditorDiv().style.borderLeft = "1px solid black";
    DOMUtils.getEditorDiv().style.paddingLeft = "10px";
    DOMUtils.getEditorDiv().style.paddingRight = "10px";
  }

  static closeEditor(): void {
    this.getToggleEditorButton().textContent = "ðŸ–‰ Open Editor";
    DOMUtils.getEditorDiv().style.width = "0";
    DOMUtils.getEditorDiv().style.borderLeft = "";
    DOMUtils.getEditorDiv().style.paddingLeft = "0";
    DOMUtils.getEditorDiv().style.paddingRight = "0";
  }

  static removeLoadModalDialog(): void {
    document.getElementById("load_modal_dialog").remove();
  }
}

export { DOMUtils };
