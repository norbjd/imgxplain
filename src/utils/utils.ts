class Utils {
  static isEquivalent(a: unknown, b: unknown): boolean {
    if (typeof a != typeof b) {
      return false;
    }

    const aProps = Object.getOwnPropertyNames(a);
    const bProps = Object.getOwnPropertyNames(b);

    if (aProps.length != bProps.length) {
      return false;
    }

    for (let i = 0; i < aProps.length; i++) {
      const propName = aProps[i];
      if (a[propName] !== b[propName]) {
        return false;
      }
    }

    return true;
  }
}

export { Utils };
