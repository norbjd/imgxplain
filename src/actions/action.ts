import { Canvas } from "../canvas";

class Action<T extends ActionOptions> {
  canvas: Canvas;
  type: string;
  options: T;

  constructor(canvas: Canvas, type: string, options: T) {
    this.canvas = canvas;
    this.type = type;
    this.options = options;
  }

  do(context: CanvasRenderingContext2D): void {
    throw new Error("Not implemented");
  }

  execute(): void { this.do(this.canvas.mainContext); }
  preview(): void { this.do(this.canvas.previewContext); }
}

class ActionOptions {}

export {
  Action,
  ActionOptions,
};
