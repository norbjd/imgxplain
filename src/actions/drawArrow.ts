import { Action, ActionOptions } from "./action";
import { Canvas } from "../canvas";

class DrawArrowOptions extends ActionOptions {
  originX: number;
  originY: number;
  destinationX: number;
  destinationY: number;
  lineWidth: number;
  color: string;

  constructor(
    originX: number,
    originY: number,
    destinationX: number,
    destinationY: number,
    lineWidth: number,
    color: string
  ) {
    super();
    this.originX = originX;
    this.originY = originY;
    this.destinationX = destinationX;
    this.destinationY = destinationY;
    this.lineWidth = lineWidth;
    this.color = color;
  }
}

class DrawArrow extends Action<DrawArrowOptions> {
  constructor(canvas: Canvas, options: DrawArrowOptions) {
    super(canvas, "draw_arrow", options);
  }

  do(context: CanvasRenderingContext2D): void {
    const lineCanvasOrigin = this.canvas.getPointProjectionOnCanvas(
      this.options.originX,
      this.options.originY
    );
    const lineCanvasOriginX = lineCanvasOrigin.x;
    const lineCanvasOriginY = lineCanvasOrigin.y;

    const lineCanvasDestination = this.canvas.getPointProjectionOnCanvas(
      this.options.destinationX,
      this.options.destinationY
    );
    const lineCanvasDestinationX = lineCanvasDestination.x;
    const lineCanvasDestinationY = lineCanvasDestination.y;

    context.save();

    context.lineWidth = this.options.lineWidth;
    context.strokeStyle = this.options.color;

    context.beginPath();
    context.moveTo(lineCanvasOriginX, lineCanvasOriginY);
    context.lineTo(lineCanvasDestinationX, lineCanvasDestinationY);
    context.stroke();
    context.closePath();

    context.beginPath();

    const arrowHeadRadius = this.options.lineWidth * 2;

    let angle = Math.atan2(
      lineCanvasDestinationY - lineCanvasOriginY,
      lineCanvasDestinationX - lineCanvasOriginX
    );
    context.moveTo(
      arrowHeadRadius * Math.cos(angle) + lineCanvasDestinationX,
      arrowHeadRadius * Math.sin(angle) + lineCanvasDestinationY
    );
    angle += (1.0 / 3.0) * (2 * Math.PI);
    context.lineTo(
      arrowHeadRadius * Math.cos(angle) + lineCanvasDestinationX,
      arrowHeadRadius * Math.sin(angle) + lineCanvasDestinationY
    );
    angle += (1.0 / 3.0) * (2 * Math.PI);
    context.lineTo(
      arrowHeadRadius * Math.cos(angle) + lineCanvasDestinationX,
      arrowHeadRadius * Math.sin(angle) + lineCanvasDestinationY
    );
    context.closePath();

    context.fillStyle = this.options.color;
    context.fill();

    context.restore();
  }
}

export { DrawArrow, DrawArrowOptions };
