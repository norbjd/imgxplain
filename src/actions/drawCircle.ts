import { Action, ActionOptions } from "./action";
import { Canvas } from "../canvas";

class DrawCircleOptions extends ActionOptions {
  centerX: number;
  centerY: number;
  radius: number;
  borderSize: number;
  color: string;

  constructor(
    centerX: number,
    centerY: number,
    radius: number,
    borderSize: number,
    color: string
  ) {
    super();
    this.centerX = centerX;
    this.centerY = centerY;
    this.radius = radius;
    this.borderSize = borderSize;
    this.color = color;
  }
}

class DrawCircle extends Action<DrawCircleOptions> {
  constructor(canvas: Canvas, options: DrawCircleOptions) {
    super(canvas, "draw_circle", options);
  }

  do(context: CanvasRenderingContext2D): void {
    const circleCanvasCenter = this.canvas.getPointProjectionOnCanvas(
      this.options.centerX,
      this.options.centerY
    );
    const circleCanvasCenterX = circleCanvasCenter.x;
    const circleCanvasCenterY = circleCanvasCenter.y;

    const ratio = this.canvas.getImageRatio();
    const circleCanvasRadiusX = this.options.radius / ratio.ratioWidth;
    const circleCanvasRadiusY = this.options.radius / ratio.ratioHeight;

    context.save();

    context.beginPath();

    context.lineWidth = this.options.borderSize;
    context.strokeStyle = this.options.color;
    context.arc(
      circleCanvasCenterX,
      circleCanvasCenterY,
      circleCanvasRadiusX,
      0,
      Math.PI
    );
    context.stroke();
    context.beginPath();
    context.arc(
      circleCanvasCenterX,
      circleCanvasCenterY,
      circleCanvasRadiusY,
      Math.PI,
      2 * Math.PI
    );
    context.stroke();

    context.restore();
  }
}

export { DrawCircle, DrawCircleOptions };
