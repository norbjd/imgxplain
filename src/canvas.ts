import { Action, ActionOptions } from "./actions/action";
import { Zoom, ZoomOptions } from "./actions/zoom";
import { DOMUtils } from "./utils/domUtils";

class Canvas {
  img: HTMLImageElement;
  imgTopLeftX: number;
  imgTopLeftY: number;
  imgBottomRightX: number;
  imgBottomRightY: number;

  mainCanvas: HTMLCanvasElement;
  mainContext: CanvasRenderingContext2D;

  previewCanvas: HTMLCanvasElement;
  previewContext: CanvasRenderingContext2D;

  clickPosition: { x: number; y: number };

  ccEnabled: boolean;

  constructor(imgBlob: string) {
    this.img = new Image();
    this.img.src = imgBlob;
    this.img.addEventListener("load", () => {
      this.renderImage();
    });

    this.imgTopLeftX = 0;
    this.imgTopLeftY = 0;
    this.imgBottomRightX = this.img.naturalWidth;
    this.imgBottomRightY = this.img.naturalHeight;

    this.mainCanvas = DOMUtils.getMainCanvas();
    this.mainContext = this.mainCanvas.getContext("2d");

    this.previewCanvas = DOMUtils.getPreviewCanvas();
    this.previewContext = this.previewCanvas.getContext("2d");

    this.clickPosition = { x: 0, y: 0 };

    this.ccEnabled = false;
  }

  render(
    topLeftX: number,
    topLeftY: number,
    bottomRightX: number,
    bottomRightY: number
  ): void {
    this.imgTopLeftX = topLeftX;
    this.imgTopLeftY = topLeftY;
    this.imgBottomRightX = bottomRightX;
    this.imgBottomRightY = bottomRightY;

    const imgWidth = this.imgBottomRightX - this.imgTopLeftX;
    const imgHeight = this.imgBottomRightY - this.imgTopLeftY;

    const imgRatio = imgWidth / imgHeight;

    const mainDivMaxWidth = DOMUtils.getAvailableMaxWidthForMainDiv();
    const mainDivMaxHeight = DOMUtils.getAvailableMaxHeightForMainDiv();

    let canvasWidth = Math.min(mainDivMaxWidth, imgWidth);
    let canvasHeight = canvasWidth / imgRatio;

    if (canvasHeight >= mainDivMaxHeight) {
      canvasHeight = mainDivMaxHeight;
      canvasWidth = canvasHeight * imgRatio;
    }

    this.mainCanvas.width = canvasWidth;
    this.mainCanvas.height = canvasHeight;
    this.mainContext.drawImage(
      this.img,
      this.imgTopLeftX,
      this.imgTopLeftY,
      imgWidth,
      imgHeight,
      0,
      0,
      canvasWidth,
      canvasHeight
    );

    this.previewCanvas.width = canvasWidth;
    this.previewCanvas.height = canvasHeight;
    this.previewCanvas.style.display = "none";

    // TODO : move somewhere else
    // -20 : padding
    DOMUtils.getTranscriptDiv().style.width = "" + (canvasWidth + 5 - 20);
  }

  renderFullImage(): void {
    this.render(0, 0, this.img.naturalWidth, this.img.naturalHeight);
  }

  renderImage(): void {
    this.render(
      this.imgTopLeftX,
      this.imgTopLeftY,
      this.imgBottomRightX,
      this.imgBottomRightY
    );
  }

  getWidthAndHeight(): { width: number; height: number } {
    return {
      width: this.mainCanvas.width,
      height: this.mainCanvas.height,
    };
  }

  getImageRatio(): { ratioWidth: number; ratioHeight: number } {
    const imgWidth = this.imgBottomRightX - this.imgTopLeftX;
    const imgHeight = this.imgBottomRightY - this.imgTopLeftY;

    const canvasWidthAndHeight = this.getWidthAndHeight();

    return {
      ratioWidth: imgWidth / canvasWidthAndHeight.width,
      ratioHeight: imgHeight / canvasWidthAndHeight.height,
    };
  }

  getPointProjectionOnCanvas(
    imgPointX: number,
    imgPointY: number
  ): { x: number; y: number } {
    const ratio = this.getImageRatio();

    const canvasPointX = (imgPointX - this.imgTopLeftX) / ratio.ratioWidth;
    const canvasPointY = (imgPointY - this.imgTopLeftY) / ratio.ratioHeight;

    return {
      x: canvasPointX,
      y: canvasPointY,
    };
  }

  getCursorPositionOnCanvas(event: MouseEvent): { x: number; y: number } {
    const rect = this.mainCanvas.getBoundingClientRect();
    const x = event.clientX - rect.left;
    const y = event.clientY - rect.top;

    const ratio = this.getImageRatio();

    return {
      x: Math.floor(x * ratio.ratioWidth) + this.imgTopLeftX,
      y: Math.floor(y * ratio.ratioHeight) + this.imgTopLeftY,
    };
  }

  addClickEventListener(listener: (event: MouseEvent) => void): void {
    this.mainCanvas.addEventListener("click", listener);
  }

  executeAction(action: Action<ActionOptions>): void {
    action.execute();
  }

  executeActions(actions: Action<ActionOptions>[]): void {
    if (actions.find((action) => action instanceof Zoom) == undefined) {
      this.renderFullImage();
    }

    this.renderImage();
    actions.forEach((action) => {
      this.executeAction(action);
    });
  }

  previewAction(action: Action<ActionOptions>, timeInMs: number): void {
    this.previewCanvas.style.display = "block";
    action.preview();
    setTimeout(() => {
      this.previewContext.clearRect(
        0,
        0,
        this.previewCanvas.width,
        this.previewCanvas.height
      );
      this.previewCanvas.style.display = "none";
      if (action instanceof Zoom) {
        const unzoom = new Zoom(
          this,
          new ZoomOptions(0, 0, this.img.naturalWidth, this.img.naturalHeight)
        );
        this.executeAction(unzoom);
      }
    }, timeInMs);
  }
}

export { Canvas };
