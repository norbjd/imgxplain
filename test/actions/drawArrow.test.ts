import { expect, it, jest, beforeEach } from "@jest/globals";
import { DrawArrow, DrawArrowOptions } from "../../src/actions/drawArrow";
import { Canvas } from "../../src/canvas";
import { DOMUtils } from "../../src/utils/domUtils";

import { readFileSync } from "fs";
import { resolve } from "path";
import { initCanvas } from "./utils";

const html = readFileSync(
  resolve(__dirname, "../../public/index.html"),
  "utf8"
);
jest.dontMock("fs");
document.documentElement.innerHTML = html.toString();

let canvas: Canvas;
beforeEach(() => {
  canvas = initCanvas();
});

it("should draw a black arrow", () => {
  new DrawArrow(
    canvas,
    new DrawArrowOptions(10, 40, 250, 130, 3, "#000000")
  ).execute();
  const imgBase64 = DOMUtils.getMainCanvas().toDataURL("image/png");
  expect(imgBase64).toMatchSnapshot();
});

it("should preview a black arrow", () => {
  new DrawArrow(
    canvas,
    new DrawArrowOptions(10, 40, 250, 130, 3, "#000000")
  ).preview();
  const imgBase64 = DOMUtils.getPreviewCanvas().toDataURL("image/png");
  expect(imgBase64).toMatchSnapshot();
});
