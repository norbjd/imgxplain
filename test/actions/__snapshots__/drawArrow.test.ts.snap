// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should draw a black arrow 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAAEc0lEQVR4nO3Yvy40fRjH4d8jEYnmqRScARXFOgJKiUa7QukYHIXQy1ZCnIJOLypKUWyHQjSS3ad5933Hy8P+mdmZe+a6yt1J9q4+me/+6vf7/QQQwEzZBwAMS7CAMAQLCEOwgDAECwhDsIAwBAsIQ7CAMAQLCEOwgDAECwhDsIAwBAsIQ7CAMAQLCEOwgDAECwhDsIAwBAsIQ7CAMAQLCEOwgDAECwhDsIAwBAsIQ7CAMAQLCEOwgDAECwhDsIAwBAsIQ7CAMAQLCEOwgDAECwhDsIAwBAsIY+hg3d7epqenpyJvAfjW0MHa3d1NS0tLaWtrK11cXKT39/ci7wL45Fe/3+//9NDNzU1aW1v78Nni4mLa2dlJe3t7aXV1tbADAQaGesN6fX1NrVbrw2fdbjcdHR2ltbW1tL6+nk5OTkxGoFBDvWEN3N3dpbOzs3R6epoeHh4+fT83N5c2NzdTu91O29vbaXZ2NtdjgWYbKVgDvV4vXV1dpU6nky4vL9Pb29unZ0xGIG9jBSvr5eUlnZ+fp06nk66vr798ZmVlJbXb7bS/v58WFhYm+TmgwSYOVpbJCBQp12ANmIxAEQoJVpbJCOSl8GBlmYzAJKYarAGTERhHKcHKMhmBYZUerCyTEfhOpYI1YDICX6lksLJMRmCg8sHKMhmh2UIFa8BkhGYKGawskxGaI3ywskxGqLdaBWvAZIR6qmWwskxGqI/aByvLZITYGhWsAZMRYmpksLJMRoij8cHKMhmh2gTrCyYjVJNg/cBkhOoQrBGYjFAuwRqDyQjlEKwJmYwwPYKVI5MRiiVYBTAZoRiCVTCTEfIjWFNkMsJkBKsEJiOMR7BKZjLC8ASrQkxG+J5gVZDJCF8TrIozGeE/ghWIyUjTCVZAJiNNJVjBmYw0iWDViMlI3QlWDZmM1JVg1ZzJSJ0IVoOYjEQnWA1kMhKVYDWcyUgkgsW/TEaqTrD4xGSkqgSLb5mMVIlgMTSTkbIJFiMzGSmLYDERk5FpEixyYzJSNMEidyYjRREsCmUykifBYmpMRiYlWEydyci4BItSmYyMQrCoDJORnwgWlWMy8jeCRaWZjGQJFmGYjAgW4ZiMzSVYhGYyNotgURsmY/0JFrVjMtaXYFFrJmO9CBaNYTLGJ1g0jskYl2DRaCZjLIIF/zAZq0+w4H9MxuoSLPiGyVgtggVDMhnLJ1gwIpOxPIIFEzAZp0uwICcmY/EEC3JmMhZHsKBAJmO+BAumxGScnGDBlJmM4xMsKJHJOBrBgoowGX8mWFAxJuPfCRZUmMn4kWBBECajYEE4w07Gw8PDdHBwUMKFxZkp+wBgNDMzM2ljYyN1Op30+PiYjo+PU6vV+vBMt9tNv3//LunC4njDgprITsbn5+fU7XbT/Px82WflSrCgZnq9Xrq/v0/Ly8tln5I7wQLC8B8WEIZgAWEIFhCGYAFhCBYQhmABYQgWEIZgAWEIFhCGYAFhCBYQhmABYQgWEIZgAWEIFhCGYAFhCBYQhmABYfwBWNIHhrZLrpkAAAAASUVORK5CYII="`;

exports[`should preview a black arrow 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAAEc0lEQVR4nO3Yvy40fRjH4d8jEYnmqRScARXFOgJKiUa7QukYHIXQy1ZCnIJOLypKUWyHQjSS3ad5933Hy8P+mdmZe+a6yt1J9q4+me/+6vf7/QQQwEzZBwAMS7CAMAQLCEOwgDAECwhDsIAwBAsIQ7CAMAQLCEOwgDAECwhDsIAwBAsIQ7CAMAQLCEOwgDAECwhDsIAwBAsIQ7CAMAQLCEOwgDAECwhDsIAwBAsIQ7CAMAQLCEOwgDAECwhDsIAwBAsIQ7CAMAQLCEOwgDAECwhDsIAwBAsIY+hg3d7epqenpyJvAfjW0MHa3d1NS0tLaWtrK11cXKT39/ci7wL45Fe/3+//9NDNzU1aW1v78Nni4mLa2dlJe3t7aXV1tbADAQaGesN6fX1NrVbrw2fdbjcdHR2ltbW1tL6+nk5OTkxGoFBDvWEN3N3dpbOzs3R6epoeHh4+fT83N5c2NzdTu91O29vbaXZ2NtdjgWYbKVgDvV4vXV1dpU6nky4vL9Pb29unZ0xGIG9jBSvr5eUlnZ+fp06nk66vr798ZmVlJbXb7bS/v58WFhYm+TmgwSYOVpbJCBQp12ANmIxAEQoJVpbJCOSl8GBlmYzAJKYarAGTERhHKcHKMhmBYZUerCyTEfhOpYI1YDICX6lksLJMRmCg8sHKMhmh2UIFa8BkhGYKGawskxGaI3ywskxGqLdaBWvAZIR6qmWwskxGqI/aByvLZITYGhWsAZMRYmpksLJMRoij8cHKMhmh2gTrCyYjVJNg/cBkhOoQrBGYjFAuwRqDyQjlEKwJmYwwPYKVI5MRiiVYBTAZoRiCVTCTEfIjWFNkMsJkBKsEJiOMR7BKZjLC8ASrQkxG+J5gVZDJCF8TrIozGeE/ghWIyUjTCVZAJiNNJVjBmYw0iWDViMlI3QlWDZmM1JVg1ZzJSJ0IVoOYjEQnWA1kMhKVYDWcyUgkgsW/TEaqTrD4xGSkqgSLb5mMVIlgMTSTkbIJFiMzGSmLYDERk5FpEixyYzJSNMEidyYjRREsCmUykifBYmpMRiYlWEydyci4BItSmYyMQrCoDJORnwgWlWMy8jeCRaWZjGQJFmGYjAgW4ZiMzSVYhGYyNotgURsmY/0JFrVjMtaXYFFrJmO9CBaNYTLGJ1g0jskYl2DRaCZjLIIF/zAZq0+w4H9MxuoSLPiGyVgtggVDMhnLJ1gwIpOxPIIFEzAZp0uwICcmY/EEC3JmMhZHsKBAJmO+BAumxGScnGDBlJmM4xMsKJHJOBrBgoowGX8mWFAxJuPfCRZUmMn4kWBBECajYEE4w07Gw8PDdHBwUMKFxZkp+wBgNDMzM2ljYyN1Op30+PiYjo+PU6vV+vBMt9tNv3//LunC4njDgprITsbn5+fU7XbT/Px82WflSrCgZnq9Xrq/v0/Ly8tln5I7wQLC8B8WEIZgAWEIFhCGYAFhCBYQhmABYQgWEIZgAWEIFhCGYAFhCBYQhmABYQgWEIZgAWEIFhCGYAFhCBYQhmABYfwBWNIHhrZLrpkAAAAASUVORK5CYII="`;
